include/aspect/melt.h

```diff
@@ -0,0 +1,329 @@
+         * required for compressible models to describe volume changes
+         * of melt in dependence of pressure, temperature etc.
+         */
+        std::vector<Tensor<1,dim> > fluid_density_gradients;
+
+        void average (const MaterialAveraging::AveragingOperation operation,
+                      const FullMatrix<double>  &projection_matrix,
+                      const FullMatrix<double>  &expansion_matrix)
+        {
+          average_property (operation, projection_matrix, expansion_matrix,
+                            compaction_viscosities);
+          average_property (operation, projection_matrix, expansion_matrix,
+                            fluid_viscosities);
+          average_property (operation, projection_matrix, expansion_matrix,
+                            permeabilities);
+          average_property (operation, projection_matrix, expansion_matrix,
+                            fluid_densities);
+
+          // The fluid density gradients are unfortunately stored in reverse
+          // indexing and averaging is not implemented for tensors (only for
+          // doubles). It's also not quite clear whether these should
+          // really be averaged, so avoid this for now
+        }
+    };
+
+    /**
+     * Base model for material models that implement a melt fraction function.
+     * This is used to compute some statistics about the melt fraction.
+     */
+    template <int dim>
+    class MeltFractionModel
+    {
+      public:
+        virtual void melt_fractions (const MaterialModel::MaterialModelInputs<dim> &in,
+                                     std::vector<double> &melt_fractions) const = 0;
+    };
+
+  }
+
+  /**
+   * Creates additional material model output objects that are
+   * needed for a simulation, and attaches a pointer to them to the
+   * corresponding vector in the MaterialModel::MaterialModelOutputs
+   * structure.
+   */
+  template <int dim>
+  void create_melt_material_outputs(MaterialModel::MaterialModelOutputs<dim> &output);
+
+
+  namespace Assemblers
+  {
+    /**
+     * A class for the definition of functions that implement the
+     * linear system terms for the *melt* migration compressible or
+     * incompressible equations.
+     */
+    template <int dim>
+    class MeltEquations : public aspect::internal::Assembly::Assemblers::AssemblerBase<dim>,
+      public SimulatorAccess<dim>
+    {
+      public:
+        /**
+         * Attaches melt outputs.
+         */
+        virtual
+        void
+        create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &outputs) const;
+
+
+        /**
+         * Compute the integrals for the preconditioner for the Stokes system in
+         * the case of melt migration on a single cell.
+         */
+        void
+        local_assemble_stokes_preconditioner_melt (const double                                             pressure_scaling,
```

Comment:
Here and in the functions below, it does not actually matter how you name them, but `_melt` is already somehow implied by the name of the class. It's not an important point to me, so I leave it to you whether or not you want to keep that part of the name.